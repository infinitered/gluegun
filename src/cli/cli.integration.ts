import * as expect from 'expect'
import * as sinon from 'sinon'
import * as uniqueTempDir from 'unique-temp-dir'
import * as stripANSI from 'strip-ansi'
import { run as cli } from './cli'

sinon.stub(console, 'log')

const pwd = process.cwd()

// set jest timeout to very long, because these take a while
beforeAll(() => jest.setTimeout(60 * 1000))
// reset back
afterAll(() => jest.setTimeout(5 * 1000))

test('can start the cli', async () => {
  const c = await cli()
  expect(c).toBeTruthy()
})

test('can create a new boilerplate cli', async () => {
  const tmp = uniqueTempDir({ create: true })
  process.chdir(tmp as string)

  const toolbox = await cli('new foo')
  expect(toolbox.command.name).toBe('new')

  const pkg = toolbox.filesystem.read(`${tmp}/foo/package.json`, 'json')

  expect(typeof pkg).toBe('object')
  expect(pkg.name).toBe('foo')
  expect(pkg.private).toBeTruthy()
  expect(Object.keys(pkg.dependencies).includes('gluegun')).toBeTruthy()

  // Install local version of gluegun to test
  await toolbox.system.run(`cd ${tmp}/foo && npm install ${pwd}`)

  // Try running the help command, see what it does
  const runCommand = await toolbox.system.run(`${tmp}/foo/bin/foo --help`)
  expect(stripANSI(runCommand)).toMatchSnapshot()

  // Try running the generate command, see what it does
  const genCommand = await toolbox.system.run(`${tmp}/foo/bin/foo g model test`)
  expect(stripANSI(genCommand)).toMatchSnapshot()

  // clean up
  process.chdir(pwd)
  toolbox.filesystem.remove(`${tmp}/foo`)
})

test('can create a new boilerplate TypeScript cli', async () => {
  const tmp = uniqueTempDir({ create: true })
  process.chdir(tmp as string)

  const toolbox = await cli('new foo-ts --typescript')
  expect(toolbox.command.name).toBe('new')

  const pkg = toolbox.filesystem.read(`${tmp}/foo-ts/package.json`, 'json')

  expect(typeof pkg).toBe('object')
  expect(pkg.name).toBe('foo-ts')
  expect(pkg.private).toBeTruthy()
  expect(Object.keys(pkg.dependencies).includes('gluegun')).toBeTruthy()

  // Install local version of gluegun to test
  await toolbox.system.run(`cd ${tmp}/foo-ts && npm install ${pwd}`)

  // Try running the help command, see what it does
  const runCommand = await toolbox.system.run(`${tmp}/foo-ts/bin/foo-ts --help`)
  expect(stripANSI(runCommand)).toMatchSnapshot()

  // Try running the generate command, see what it does
  const genCommand = await toolbox.system.run(`${tmp}/foo-ts/bin/foo-ts g model test`)
  expect(stripANSI(genCommand)).toMatchSnapshot()

  // Add a command that exercises a lot of Gluegun features
  // Incidentally, it verifies that the template tool works
  const generateResult = await toolbox.template.generate({
    template: `test/kitchen-sink-command.js.ejs`,
    target: `${tmp}/foo-ts/commands/kitchen.js`,
  })

  // Verify the result of the generated command
  expect(generateResult.includes('module.exports = {')).toBe(true)

  // Run that command and check the result
  const kitchenCommand = await toolbox.system.run(`${tmp}/foo-ts/bin/foo-ts kitchen`)
  console.log(kitchenCommand)

  // clean up
  process.chdir(pwd)
  toolbox.filesystem.remove(`${tmp}/foo-ts`)
})
