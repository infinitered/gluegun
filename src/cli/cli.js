const { build } = require('../index.js')

/**
 * Create the cli and kick it off
 */
async function run (argv) {
  // create a CLI runtime
  const cli = build()
    .brand('gluegun')
    .src(`${__dirname}`)
    .help()
    .version()
    .create()

  // and run it
  const context = await cli.run(argv)

  // send it back (for testing, mostly)
  return context
}

module.exports = run
