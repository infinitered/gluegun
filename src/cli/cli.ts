import { build, GluegunToolbox } from '../index'

/**
 * Create the cli and kick it off
 */
export async function run(argv?: string[] | string): Promise<GluegunToolbox> {
  // create a CLI runtime
  const gluegunCLI = build('gluegun')
    .src(__dirname)
    .help()
    .version()
    .defaultCommand({
      run: async (toolbox: GluegunToolbox) => {
        const { print, meta } = toolbox
        print.info(`Gluegun version ${meta.version()}`)
        print.info(``)
        print.info(`  Type gluegun --help for more info`)
      },
    })
    .exclude(['http', 'patching'])
    .checkForUpdates(25)
    .create()

  // and run it
  const toolbox = await gluegunCLI.run(argv)

  // send it back (for testing, mostly)
  return toolbox
}
