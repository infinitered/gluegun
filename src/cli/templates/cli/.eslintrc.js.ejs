/**
 * @type {import("eslint").Linter.Config}
 */
module.exports = {
  <% if (props.language === "typescript") { %>
    parser: '@typescript-eslint/parser',
  <% } %>
  parserOptions: {
    ecmaVersion: 2020,
    sourceType: 'module'
  },
  extends: [
  <% if (props.language === "typescript") { %>
    'plugin:@typescript-eslint/recommended',
  <% } %>
    'prettier',
    'plugin:prettier/recommended'
  ],
  rules: {}
}
