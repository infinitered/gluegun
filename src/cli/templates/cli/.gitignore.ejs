.DS_Store
node_modules
npm-debug.log
coverage
.nyc_output
dist
build
