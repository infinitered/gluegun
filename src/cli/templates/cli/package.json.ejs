{
  "name": "<%= props.name %>",
  "version": "0.0.1",
  "description": "<%= props.name %> CLI",
  "private": true,
  <% if (props.language === "typescript") { %>
    "types": "build/types/types.d.ts",
  <% } %>
  "bin": {
    "<%= props.name %>": "bin/<%= props.name %>"
  },
  "scripts": {
    <% if (props.language === "typescript") { %>
      "clean-build": "rm -rf ./build",
      "compile": "tsc -p .",
      "copy-templates": "copyfiles ./src/templates/* ./build/templates",
      "build": "yarn clean-build && yarn compile && yarn copy-templates",
      "prepublishOnly": "yarn build",
      "format": "eslint \"**/*.{js,jsx,ts,tsx}\" --fix && prettier \"**/*.{js,jsx,ts,tsx,json}\" --write",
    <% } else { %>
      "format": "eslint \"**/*.{js,jsx}\" --fix && prettier \"**/*.{js,jsx,json}\" --write",
    <% } %>
    "test": "jest",
    "watch": "jest --watch",
    "snapupdate": "jest --updateSnapshot",
    "coverage": "jest --coverage"
  },
  "files": [
    <% if (props.language === "typescript") { %>
      "build",
    <% } else { %>
      "src",
    <% } %>
    "LICENSE",
    "readme.md",
    "docs",
    "bin"
  ],
  "license": "MIT",
  "dependencies": {
    "gluegun": "latest"
  },
  "devDependencies": {
    <% if (props.language === "typescript") { %>
      "@types/node": "^12.7.11",
      "@types/jest": "^26.0.20",
      "@typescript-eslint/eslint-plugin": "^4.17.0",
      "@typescript-eslint/parser": "^4.17.0",
      "ts-jest": "^26.5.3",
      "ts-node": "^9.1.1",
      "typescript": "^4.2.3",
    <% } %>
    "copyfiles": "^2.4.1",
    "eslint": "^7.22.0",
    "eslint-config-prettier": "^8.1.0",
    "eslint-plugin-prettier": "^3.3.1",
    "husky": "^5.1.3",
    "jest": "^26.6.3",
    "prettier": "^2.2.1",
    "pretty-quick": "^3.1.0"
  },
  "jest": {
    <% if (props.language === "typescript") { %>
      "preset": "ts-jest",
    <% } %>
    "testEnvironment": "node"
  },
  "prettier": {
    "semi": false,
    "singleQuote": true
  },
  "husky": {
    "hooks": {
      "pre-commit": "pretty-quick --staged"
    }
  }
}
