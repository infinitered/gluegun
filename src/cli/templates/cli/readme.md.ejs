# <%= props.name %> CLI

A CLI for <%= props.name %>.

## Customizing your CLI

Check out the documentation at https://github.com/infinitered/gluegun/tree/master/docs.

## Publishing to NPM

To package your CLI up for NPM, do this:

```shell
$ npm login
$ npm whoami
$ npm test
<% if (props.language === "typescript") { %>
$ npm run build
<% } %>
$ npm publish
```

# License

MIT - see LICENSE

