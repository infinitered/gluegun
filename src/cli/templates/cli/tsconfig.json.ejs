{
  "compilerOptions": {
    "allowSyntheticDefaultImports": false,
    "experimentalDecorators": true,
    "lib": ["es2015", "scripthost", "es2015.promise", "es2015.generator", "es2015.iterable", "dom"],
    "module": "commonjs",
    "moduleResolution": "node",
    "noImplicitAny": false,
    "noImplicitThis": true,
    "noUnusedLocals": true,
    "sourceMap": true,
    "inlineSourceMap": true,
    "outDir": "build",
    "strict": false,
    "target": "es5",
    "declaration": true,
    "declarationDir": "build/types",
  },
  "include": ["src/**/*"],
  "exclude": ["node_modules"]
}
