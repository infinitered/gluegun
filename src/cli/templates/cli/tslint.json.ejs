{
  "extends": ["tslint-config-standard", "tslint-config-prettier"],
  "rules": {
    "strict-type-predicates": false
  },
  "env": {
    "jest": true
  }
}
