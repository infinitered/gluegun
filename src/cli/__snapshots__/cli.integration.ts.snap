// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can create a new boilerplate TypeScript cli 1`] = `
"[0mfoo-ts version 0.0.1[0m

  foo-ts         -                         
  generate (g)   -                         
  help (h)       -                         
  version (v)    Output the version number 
"
`;

exports[`can create a new boilerplate TypeScript cli 2`] = `""`;

exports[`can create a new boilerplate cli 1`] = `
"[0mfoo version 0.0.1[0m

  foo            -                         
  generate (g)   -                         
  help (h)       -                         
  version (v)    Output the version number 
"
`;

exports[`can create a new boilerplate cli 2`] = `""`;
